pytorch_pfn_extras/onnx/_grad.py

```diff
@@ -0,0 +1,105 @@
+from typing import Generator, List, Optional, Tuple, Any
+from contextlib import contextmanager
+
+import torch
+import torch.onnx
+import threading
+from pytorch_pfn_extras.onnx._as_output import as_output
+
+
+_grad_state = threading.local()
+
+
+@contextmanager
+def init_grad_state() -> Generator[None, None, None]:
+    _grad_state.n_grad_call = 0
+    try:
+        yield
+    finally:
+        _grad_state.n_grad_call = None
+
+
+def grad(
+    output: torch.Tensor,
+    inputs: Tuple[torch.Tensor, ...],
+    retain_graph: Optional[bool] = None,
+    create_graph: bool = False,
+    only_inputs: bool = True,
+    allow_unused: bool = False,
+) -> Tuple[Optional[torch.Tensor], ...]:
+    grad_output = torch.ones_like(output)
+
+    if torch.jit.is_tracing():  # type: ignore
+        err_msg = "ppe.onnx.grad() can only be used in conjunction " + \
+            "with export functions under ppe.onnx"
+        assert hasattr(_grad_state, "n_grad_call"), err_msg
+        n_grad_call = _grad_state.n_grad_call
+        _grad_state.n_grad_call += 1
+
+        output_name = f"Gradient_y_{n_grad_call}"
+        output = as_output(output_name, output)
+
+        assert only_inputs, "only_inputs=False case is not supported now"
+
+        input_names = []
+        inputs_l = list(inputs)
+        for i, input in enumerate(inputs):
+            input_name = f"Gradient_x_{i}_{n_grad_call}"
+            input_names.append(input_name)
+            inputs_l[i] = as_output(input_name, input)
+
+        class _Gradient(torch.autograd.Function):
+            @staticmethod
+            def forward(  # type: ignore
+                ctx: Any,
+                output: torch.Tensor,
+                grad_output: Optional[torch.Tensor],
+                *inputs: Tuple[torch.Tensor, ...],
+            ) -> Tuple[Optional[torch.Tensor], ...]:
+                @torch.jit.script
+                def _grad(
+                    output: torch.Tensor,
+                    inputs: List[torch.Tensor],
+                    grad_output: Optional[torch.Tensor],
+                    retain_graph: Optional[bool],
+                    create_graph: bool,
+                    allow_unused: bool,
+                ) -> List[Optional[torch.Tensor]]:
```

Comment:
Another mypy error :pray:
```
pytorch_pfn_extras/onnx/_grad.py:68: error: Incompatible return value type (got "Tuple[Tensor, ...]", expected "List[Optional[Tensor]]")  [return-value]
```

Reply:
hmm, the error occurs when using torch 1.8. and in torch 1.9,  using `Tuple[Tensor, ...]` causes mypy error.
I ignores the type hints in this tracing code path.

Reply:
Thanks, I agree ignoring type hints is ok here.